util/cpuprofile/cpuprofile.go

```diff
@@ -0,0 +1,243 @@
+	"github.com/pingcap/tidb/util"
+	"github.com/pingcap/tidb/util/logutil"
+	"go.uber.org/atomic"
+)
+
+var defProfileDuration = time.Second
+
+// globalCPUProfiler is the global CPU profiler.
+// If you want to create a new global cpu profiler, you should stop the old global cpu profiler.
+var globalCPUProfiler = newParallelCPUProfiler()
+
+// ProfileConsumer is the profile data consumer.
+// ProfileConsumer is a channel alias, if the channel is full, then the channel won't receive the latest profile data
+// until it is not blocked.
+type ProfileConsumer = chan *ProfileData
+
+// ProfileData contains the cpu profile data and some additional information.
+type ProfileData struct {
+	Data  *bytes.Buffer
+	Begin time.Time
+	End   time.Time
+	Error error
+}
+
+// StartCPUProfiler uses to start to run the global parallelCPUProfiler.
+func StartCPUProfiler() error {
+	return globalCPUProfiler.start()
+}
+
+// CloseCPUProfiler uses to close the global parallelCPUProfiler.
+func CloseCPUProfiler() {
+	globalCPUProfiler.close()
+}
+
+// Register register a ProfileConsumer into the global CPU profiler.
+// Normally, the registered ProfileConsumer will receive the cpu profile data per second.
+// If the ProfileConsumer (channel) is full, the latest cpu profile data will not be sent to it.
+// This function is thread-safe.
+func Register(ch ProfileConsumer) {
+	globalCPUProfiler.register(ch)
+}
+
+// Unregister unregister a ProfileConsumer from the global CPU profiler.
+// The unregistered ProfileConsumer won't receive the cpu profile data any more.
+// This function is thread-safe.
+func Unregister(ch ProfileConsumer) {
+	globalCPUProfiler.unregister(ch)
+}
+
+// ConsumersCount returns the count of the global parallelCPUProfiler's consumer.It is exporting for test.
+func ConsumersCount() int {
+	return globalCPUProfiler.consumersCount()
+}
+
+// parallelCPUProfiler is a cpu profiler.
+// With parallelCPUProfiler, it is possible to have multiple profile consumer at the same time.
+// WARN: Only one running parallelCPUProfiler is allowed in the process, otherwise some profiler may profiling fail.
+type parallelCPUProfiler struct {
+	sync.Mutex
+	cs     map[ProfileConsumer]struct{}
+	notify chan struct{}
+
+	// some data cache for profiling.
+	data         *ProfileData
+	nextData     *ProfileData
+	lastDataSize int
+
+	started  atomic.Bool
+	closed   chan struct{}
+	isClosed atomic.Bool
+	wg       sync.WaitGroup
+}
+
+// newParallelCPUProfiler crate a new parallelCPUProfiler.
+func newParallelCPUProfiler() *parallelCPUProfiler {
+	return &parallelCPUProfiler{
+		cs:     make(map[ProfileConsumer]struct{}),
+		notify: make(chan struct{}),
+		closed: make(chan struct{}),
+	}
+}
+
+var (
+	errProfilerAlreadyStarted = errors.New("parallelCPUProfiler already started")
+	errProfilerAlreadyClosed  = errors.New("parallelCPUProfiler already closed")
+)
+
+func (p *parallelCPUProfiler) start() error {
+	if !p.started.CAS(false, true) {
```

Comment:
This function is not concurrent safe, so that unfortunately your CAS means nothing.. It is even worse than a simple lock, or no concurrent guard at all.